salt/engines/libvirt_events.py

```diff
@@ -0,0 +1,742 @@
+                - domain/lifecycle
+                - domain/reboot
+                - pool
+
+The default URI is ``qemu:///system`` and the default tag prefix is
+``salt/engines/libvirt_events``.
+
+Filters is a list of event types to relay to the event bus. Items in this list
+can be either one of the main types (``domain``, ``network``, ``pool``,
+``nodedev``, ``secret``), ``all`` or a more precise filter. These can be done
+with values like <main_type>/<subtype>, but better read the engine's code for
+the complete list of sub types. If the filters list contains ``all``, all
+events will be relayed.
+
+Be aware that the list of events increases with libvirt versions, for example
+network events have been added in libvirt 1.2.1.
+
+:depends: libvirt 1.0.0+ python binding
+'''
+
+from __future__ import absolute_import
+import atexit
+import logging
+import traceback
+
+# Import salt libs
+import salt.utils.event
+
+try:
+    from urlparse import urlparse
+except ModuleNotFoundError:
+    from urllib.parse import urlparse
+
+log = logging.getLogger(__name__)
+
+
+HAS_LIBVIRT_PY = False
+try:
+    import libvirt
+    if libvirt.getVersion() >= 1000000:
+        HAS_LIBVIRT_PY = True
+    else:
+        log.error('libvirt >= 1.0.0 required')
+except ModuleNotFoundError:
+    pass
+
+__virtualname__ = 'libvirt_events'
+
+
+def __virtual__():
+    '''
+    Only load if libvirt python binding is present
+    '''
+    if not HAS_LIBVIRT_PY:
+        return (False, 'libvirt_events engine could not be imported')
+    return __virtualname__
+
+
+def nth(stack, index, default=None):
+    '''
+    returns the element at position defined by ``index`` or ``default`` if it
+    doesn't exist.
+    '''
+    try:
+        return stack[index]
+    except IndexError:
+        return default
+
+
+DOMAIN_EVENTS_MAP = (
+    ('defined', ("added",
+                 "updated",
+                 "renamed",
+                 "from snapshot")),
+    ('undefined', ("removed",
+                   "renamed")),
+    ('started', ("booted",
+                 "migrated",
+                 "restored",
+                 "snapshot",
+                 "wakeup")),
+    ('suspended', ("paused",
+                   "migrated",
+                   "ioerror",
+                   "watchdog",
+                   "restored",
+                   "snapshot",
+                   "api error",
+                   "postcopy",
+                   "postcopy failed")),
+    ('resumed', ("unpaused",
+                 "migrated",
+                 "snapshot",
+                 "postcopy")),
+    ('stopped', ("shutdown",
+                 "destroyed",
+                 "crashed",
+                 "migrated",
+                 "saved",
+                 "failed",
+                 "snapshot")),
+    ('shutdown', ("finished",
+                  "on guest request",
+                  "on host request")),
+    ('pmsuspended', ("memory",
+                     "disk")),
+    ('crashed', ("panicked"))
+)
```

Comment:
I would refactor [all this](https://github.com/cbosdo/salt/blob/5690a1a045efd5bee3f38ed8fd12025068df80fc/salt/engines/libvirt_events.py#L73-L122) into a dictionary. The `def nth` is not needed anywhere, because the dictionary has `.get(...)` function.

Reply:
Found a way to get rid of these lists by introspecting libvirt module.